File: myLeetCode/Dynamic_Programming/Basic/509. Fibonacci Number.cpp
// LeetCode 509. Fibonacci Number
// https://leetcode.com/problems/fibonacci-number/
// F(0) = 0, F(1) = 1
// F(n) = F(n - 1) + F(n - 2), for n > 1

#include <iostream>
#include <vector>

int fib(int n) {
    if (n < 0) return -1;
    std::vector<int> dp(n + 1, 0);
    dp[0] = 0;
    dp[1] = 1;
    
    for (int i = 2; i <= n; i++) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    
    return dp[n];
}


int main() {    
    std::cout << fib(4) << std::endl; // print the output: 3

    std::cout << fib(5) << std::endl; // print the output: 5

    std::cout << fib(6) << std::endl; // print the output: 8
    
    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Basic/509. Fibonacci Number.py
def fib(n):
    if n <= 1:
        return n

    dp = [0] * (n + 1)
    dp[1] = 1

    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]

    return dp[n]

# 测试用例
print(fib(0))  # 输出应该是 0
print(fib(1))  # 输出应该是 1
print(fib(2))  # 输出应该是 1
print(fib(3))  # 输出应该是 2
print(fib(4))  # 输出应该是 3
print(fib(5))  # 输出应该是 5
